### ISSUE TITLE
Contribute/Donate GPU time to CrazyAra development

### ISSUE THREAD

**Issue by Nordlandia** [QueensGambit/CrazyAra#37]
Is this available or planned to be available at a later point as of now?

**Comment by Nordlandia** [comment]
Perhaps support for CrazyHouse960 is a good idea. Beside Chess960. 

Support for 960 may be complicated. Afaik Lc0 has not started training for 960 yet. 

Do it has to be limited to what lichess support? 

Cutechess GUI support majority of the variants pychess supports. 

S-Chess or Seirawan-Chess is maybe unrealistic. 

I'm playing on https://pychess-variants.herokuapp.com/variant

**Comment by QueensGambit** [comment]
Adding 960 support isn't too difficult for this project because move generating routines and the position representation has been integrated from [_Multi-Variant Stockfish_](https://github.com/ddugovic/Stockfish).
Therefore the MCTS back-end is currently limited to (including 960 if applicable):
 * Crazyhouse
 * Atomic
 * Horde
 * King of the Hill
 * Racing Kings
 * Antichess
 * Three-Check
 * Chess960
 * Losers
 * Giveaway
 * Suicide
 * Loop
 * Extinction
 * Grid
 * Two Kings

I implemented the remaining parts to convert the neural network policy to chess960 and let it play one game vs Sjakii 1.3.1:
* https://lichess.org/NHJUIBla/white#0

The network used in this game was generated by supervised learning on chess960 lichess.org games. Generally, these networks seem to be a bit too careless about material due to the loss weighting during training.

One issue in converting an already trained network to the 960 variant is that castling is presented differently and there are no special castling moves `O-O`, `O-O-O` for the policy right now.
This means that `e1g1` and `e1c1` will be changed into `e1h1` and `e1a1` respectively.

For parsing pgn files in python and preparing supervised training data, the python chess library is used. The python chess library supports the same variants as _Multi-Variant Stockfish_ at the time of writing:
* https://python-chess.readthedocs.io/en/latest/variant.html

If Seirawan-Chess were supported, then one would need to train the network fully from selfplay and define an extended input representation with 4 additional layers to describe the position of the hawk and elephant pieces. Moreover, the MCTS back-end would need to be changed.
[Fairy-Stockfish](https://github.com/ianfab/Fairy-Stockfish) provides a more general back-end which supports more variants at the cost of speed and memory.
As a result _Fairy-Stockfish_ is 0-200 Elo weaker as _Multi-Variant Stockfish_ for the same variant:
* https://github.com/ianfab/Fairy-Stockfish/wiki/Playing-strength

_CrazyAra_ shouldn't lose as much playing strength if the _Fairy-Stockfish_ back-end were used because the main bottleneck is the neural network inference. 

I think before providing support for Seirawan-Chess it would make more sense to support traditional chess variants like Shogi.

Generally, extending _CrazyAra_ to other variants doesn't require manually fine-tuning of evaluation functions as for other engines but it is computational extensive to train a neural network and to generate selfplay games. Furthermore, if a different input and output representation of the neural network is used, then one needs to maintain different network definitions and weights which are mutually incompatible with each other. To address this problem, I proposed a unified multi variant plane representation in my master thesis which at least supports all lichess variants.

I could prepare the reinforcement learning loop for crazyhouse960 and give you instructions on how to start it. Then you can see how it develops starting from the already trained network for regular crazyhouse. After every network update it will play the previous network to measure the progress.
One important setting to define would be if all selfplay games should start from a 960 starting position or if some percentage should still use the classical starting position.
